In [94]:
from IPython.core.display import HTML
HTML('<style>code:not(.cm-s-ipython){color:#ce005c;padding:.2em 0.4em;margin:0;font-size:85%;background-color:rgba(255,0,114,0.09);border-radius:3px}</style>')

Попробуем разобраться с тем кодом на Python, который был в первой главе книги.

### Первый сниппет (кусок кода)

In [4]:
import json
from urllib.request import urlopen, HTTPError

url = "https://gdata.youtube.com/feeds/api/standardfeeds/top_rated?alt=json"

try:
    response = urlopen(url)
    contents = response.read()
    text = contents.decode('utf8')
    data = json.loads(text)

    for video in data['feed']['entry'][0:6]:
     print(video['title']['$t'])
    
except HTTPError as e:
    print(e)

HTTP Error 410: Gone


Этот код не работает, потому что книга слегка устарела и youtube выключил доступ к API. Об этом нам говорит код ответа 410. Вообще, все HTTP ответы типа 4xx — ошибки. Например, 404 — страница не найдена.

В интернетах пишут, что url https://raw.githubusercontent.com/koki0702/introducing-python/master/dummy_api/youTube_top_rated.json работает. Пробуем его.

In [10]:
import json
from urllib.request import urlopen, HTTPError

url = "https://raw.githubusercontent.com/koki0702/introducing-python/master/dummy_api/youTube_top_rated.json"

try:
    response = urlopen(url)
    contents = response.read()
    text = contents.decode('utf8')
    data = json.loads(text)

    for video in data['feed']['entry'][0:6]:
     print(video['title']['$t'])
    
except HTTPError as e:
    print(e)

Evolution of Dance - By Judson Laipply
Linkin Park - Numb
Potter Puppet Pals: The Mysterious Ticking Noise
"Chocolate Rain Original" Song by Tay Zonday
Charlie bit my finger - again !
The Mean Kitty Song


Отлично, работает. Пробуем разобраться что внутри. Смотрим на типы.

In [11]:
# response = urlopen(url)
response

Пока понятно, что это какой-то ответ (http response) сервера.

In [12]:
# contents = response.read()
contents

b'{\n    "feed": {\n\t"entry": [\n\t    {"title": {"$t": "Evolution of Dance - By Judson Laipply"}},\n\t    {"title": {"$t": "Linkin Park - Numb"}},\n\t    {"title": {"$t": "Potter Puppet Pals: The Mysterious Ticking Noise"}},\n\t    {"title": {"$t": "\\"Chocolate Rain Original\\" Song by Tay Zonday"}},\n\t    {"title": {"$t": "Charlie bit my finger - again !"}},\n\t    {"title": {"$t": "The Mean Kitty Song"}}\n\t]\n    }\n}\n'

Ура! Это содержимое файла. Видим знакомые **`\n`** (new line) и **`\t`** (tab symbol).

В самом начале строки что-то странное:

```python
b'………'
```

Это питон так обозначает бинарную строку. То есть `contents` — это пока что список байт, а не строка, и Jupyter сконвертировал нам её для удобства. Преобразуем бинарную строку к обычной строке. `utf8` — название кодировки.

In [13]:
# text = contents.decode('utf8')
text

'{\n    "feed": {\n\t"entry": [\n\t    {"title": {"$t": "Evolution of Dance - By Judson Laipply"}},\n\t    {"title": {"$t": "Linkin Park - Numb"}},\n\t    {"title": {"$t": "Potter Puppet Pals: The Mysterious Ticking Noise"}},\n\t    {"title": {"$t": "\\"Chocolate Rain Original\\" Song by Tay Zonday"}},\n\t    {"title": {"$t": "Charlie bit my finger - again !"}},\n\t    {"title": {"$t": "The Mean Kitty Song"}}\n\t]\n    }\n}\n'

То же самое, только без символа **`b`** в начале.

In [54]:
print(text.replace('    ', '  ').replace('\t', '    '))

{
  "feed": {
    "entry": [
      {"title": {"$t": "Evolution of Dance - By Judson Laipply"}},
      {"title": {"$t": "Linkin Park - Numb"}},
      {"title": {"$t": "Potter Puppet Pals: The Mysterious Ticking Noise"}},
      {"title": {"$t": "\"Chocolate Rain Original\" Song by Tay Zonday"}},
      {"title": {"$t": "Charlie bit my finger - again !"}},
      {"title": {"$t": "The Mean Kitty Song"}}
    ]
  }
}



А вот и сам JSON. По сути он очень похож на dictionary и lists в Python.
Пробуем скопипастить и вставить в переменную.

In [55]:
music = {
    "feed": {
        "entry": [
            {"title": {"$t": "Evolution of Dance - By Judson Laipply"}},
            {"title": {"$t": "Linkin Park - Numb"}},
            {"title": {"$t": "Potter Puppet Pals: The Mysterious Ticking Noise"}},
            {"title": {"$t": "\"Chocolate Rain Original\" Song by Tay Zonday"}},
            {"title": {"$t": "Charlie bit my finger - again !"}},
            {"title": {"$t": "The Mean Kitty Song"}}
        ]
    }
}

print(music)

{'feed': {'entry': [{'title': {'$t': 'Evolution of Dance - By Judson Laipply'}}, {'title': {'$t': 'Linkin Park - Numb'}}, {'title': {'$t': 'Potter Puppet Pals: The Mysterious Ticking Noise'}}, {'title': {'$t': '"Chocolate Rain Original" Song by Tay Zonday'}}, {'title': {'$t': 'Charlie bit my finger - again !'}}, {'title': {'$t': 'The Mean Kitty Song'}}]}}


Круто, конечно, что я взял и скопипастил JSON (который хранился в _text_)
и присвоил его переменной _music_, но Python умеет делать это и сам.

In [37]:
# data = json.loads(text)
data

{'feed': {'entry': [{'title': {'$t': 'Evolution of Dance - By Judson Laipply'}},
   {'title': {'$t': 'Linkin Park - Numb'}},
   {'title': {'$t': 'Potter Puppet Pals: The Mysterious Ticking Noise'}},
   {'title': {'$t': '"Chocolate Rain Original" Song by Tay Zonday'}},
   {'title': {'$t': 'Charlie bit my finger - again !'}},
   {'title': {'$t': 'The Mean Kitty Song'}}]}}

То есть **`json.loads`** берёт строку с JSONом (`text`) и пытается сконвертировать его в набор dictionary + lists + strings. 

In [38]:
type(data)

dict

In [42]:
data['feed']['entry']

[{'title': {'$t': 'Evolution of Dance - By Judson Laipply'}},
 {'title': {'$t': 'Linkin Park - Numb'}},
 {'title': {'$t': 'Potter Puppet Pals: The Mysterious Ticking Noise'}},
 {'title': {'$t': '"Chocolate Rain Original" Song by Tay Zonday'}},
 {'title': {'$t': 'Charlie bit my finger - again !'}},
 {'title': {'$t': 'The Mean Kitty Song'}}]

In [44]:
# for video in data['feed']['entry'][0:6]:
#     print(video['title']['$t'])

video = data['feed']['entry'][0]; video

{'title': {'$t': 'Evolution of Dance - By Judson Laipply'}}

In [45]:
video['title']['$t']

'Evolution of Dance - By Judson Laipply'

### Второй сниппет

In [59]:
import requests

ModuleNotFoundError: No module named 'requests'

Если у тебя здесь ошибка, всё норм. Нужно установить библиотеку requests. (Знак доллара обозначает cmd, я устанавливал от имени администратора.)

```bash
$ pip install requests
```

In [62]:
import requests

url = "https://raw.githubusercontent.com/koki0702/introducing-python/master/dummy_api/youTube_top_rated.json"

response = requests.get(url)
data = response.json()

for video in data['feed']['entry'][0:6]:
    print(video['title']['$t'])

Evolution of Dance - By Judson Laipply
Linkin Park - Numb
Potter Puppet Pals: The Mysterious Ticking Noise
"Chocolate Rain Original" Song by Tay Zonday
Charlie bit my finger - again !
The Mean Kitty Song


In [63]:
# response = requests.get(url)
response

<Response [200]>

Какой-то объект, который хранит в себе и ответ-строку, и HTTP-код ответа 200 (OK).

![](http://i.imgur.com/aylkANL.png)

In [64]:
# data = response.json()
data

{'feed': {'entry': [{'title': {'$t': 'Evolution of Dance - By Judson Laipply'}},
   {'title': {'$t': 'Linkin Park - Numb'}},
   {'title': {'$t': 'Potter Puppet Pals: The Mysterious Ticking Noise'}},
   {'title': {'$t': '"Chocolate Rain Original" Song by Tay Zonday'}},
   {'title': {'$t': 'Charlie bit my finger - again !'}},
   {'title': {'$t': 'The Mean Kitty Song'}}]}}

Уже знакомые нам данные. Остальной код такой же.

### Вывод

Пользоваться второй библиотекой проще, потому что она понятнее. Можно было даже написать всё в одну строку (разбил для чуть лучшего понимания):

In [77]:
_ = list(map(print, [
      v['title']['$t'] for v in
           requests.get(url).json()['feed']['entry']]))

Evolution of Dance - By Judson Laipply
Linkin Park - Numb
Potter Puppet Pals: The Mysterious Ticking Noise
"Chocolate Rain Original" Song by Tay Zonday
Charlie bit my finger - again !
The Mean Kitty Song


Тут я использовал генератор списков, `map` чтобы применить `print`. Для сравнения:

```python
response = requests.get(url)
data = response.json()

for video in data['feed']['entry'][0:6]:
    print(video['title']['$t'])
```

↑ мне нравится больше, проще читать и понимать.

### Задание

В интернетах я нашёл API для погоды: https://www.metaweather.com/api/. 

Описание сервиса-сайта:

> MetaWeather provides an API that delivers JSON over HTTPS for access to our data.

Что вообще даёт этот сервис:

<img src="http://i.imgur.com/ic9RlBf.png" width="400" style="margin-left:0">

Дальше идут примеры. Например, если обратиться по адресу [`https://www.metaweather.com/api/location/search/?query=london`](https://www.metaweather.com/api/location/search/?query=london), то получишь примерно такой список:

```js
[{"title":"London","location_type":"City","woeid":44418,"latt_long":"51.506321,-0.12714"}]
```

Список может быть пустым или содержать городов больше одного.

Главное в этом словаре — woeid, который можно использовать в другом запросе: [`https://www.metaweather.com/api/location/44418/`](https://www.metaweather.com/api/location/44418/).

Внутри поля `consolidated_weather` есть список вот таких словарей:

```js
{
  "id": 4736170084794368,
  "weather_state_name": "Light Rain",
  "weather_state_abbr": "lr",
  "wind_direction_compass": "WSW",
  "created": "2017-06-01T20:07:02.864040Z",
  "applicable_date": "2017-06-02",
  "min_temp": 14.395714285714286,
  "max_temp": 23.204285714285714,
  "the_temp": 21.573333333333334,
  "wind_speed": 5.497677255030622,
  "wind_direction": 241.6612799507309,
  "air_pressure": 1017.5899999999999,
  "humidity": 64,
  "visibility": 10.03576612582518,
  "predictability": 75
}
```

В нём нужно текстовое описание `weather_state_name` и `the_temp`, описывающее температуру.

**Задание**: написать скрипт (на основе requests), который будет забирать данные о погоде в Ростове или Москве и выводить их.

In [95]:
# Your code here